### PR TITLE
Add method to control the leading and trailing spaces when auto-distributing views.

### PR DESCRIPTION
**Pull request by rahul-malik**
Currently when using this method in NSArray (AutoLayout):

``` Objective-C
 - (NSArray *)autoDistributeViewsAlongAxis:(ALAxis)axis withFixedSpacing:(CGFloat)spacing alignment:(NSLayoutFormatOptions)alignment;
```

The spacing value not only specifies the spacing between the views but also a leading and trailing spacing that was unexpected when I began using it. Instead of altering this behavior, I've added a more specific method that allows you to control the leading and trailing spacing around the first and last view respectively. 

The new method I've added is:

``` Objective-C
- (NSArray *)autoDistributeViewsAlongAxis:(ALAxis)axis withFixedSpacing:(CGFloat)spacing withLeadingSpacing:(CGFloat)leadingSpacing withTrailingSpacing:(CGFloat)trailingSpacing alignment:(NSLayoutFormatOptions)alignment;
```

### CODE REVIEW

**Comment by smileyborg** [comment]
I see exactly what you're talking about. The pull request/code looks solid.

A couple questions:
- What about the other variant which uses a fixed size for each of the views, wouldn't this also apply to that? (The math on that one is trickier and may not be self explanatory, btw.)
- Perhaps instead of exposing specific values for leading/trailing spacing, we just add an optional parameter that is a BOOL, which if YES includes the spacing before & after, but if NO the leading/trailing spacing is zero? In other words, is there really any reason to have to pass 3 different CGFloat values, since it seems there are only 2 probable uses?

e.g. something like:

```
(NSArray *)autoDistributeViewsAlongAxis:(ALAxis)axis withFixedSpacing:(CGFloat)spacing insetSpacing:(BOOL)shouldSpaceInsets alignment:(NSLayoutFormatOptions)alignment;
```

**Comment by rahul-malik** [comment]
> What about the other variant which uses a fixed size for each of the views, wouldn't this also apply to that? (The math on that one is trickier and may not be self explanatory, btw.)
- I briefly looked at this and I'd be willing to make these modifications but the math that is currently used is a bit difficult to decipher. If you can clarify how the multiplier and constant values are calculated I can try to update it as well.

``` Objective-C
CGFloat multiplier = (i * 2.0f + 2.0f) / (numberOfViews + 1.0f);
CGFloat constant = (multiplier - 1.0f) * size / 2.0f;
```

> Perhaps instead of exposing specific values for leading/trailing spacing, we just add an optional parameter that is a BOOL, which if YES includes the spacing before & after, but if NO the leading/trailing spacing is zero? In other words, is there really any reason to have to pass 3 different CGFloat values, since it seems there are only 2 probable uses?
- I think the flexibility could allow for new use cases. I could imagine situations where you want to just apply a particular leading spacing and distribute the views amongst the remaining space. I can also add the method you've specified (or update the existing one) with the BOOL approach which would then call the method I added with zero trailing and leading spacing to simply that use case.

Thoughts?

**Comment by smileyborg** [comment]
I can take a stab at this when I get some spare time (probably this weekend), it's probably more difficult to explain the math behind that other method without being in person :P

As far as what to expose, this comes back to the core design philosophy behind the API. The goal is not to solve everything, but instead to solve most things very well, with as thin & simple of a front end as possible.

For this specific case, given that you must already have a container view to distribute these subviews within, I'm going to say that it's therefore very easy to adjust the leading and trailing (or top & bottom) spacing to the first/last views by positioning the superview (assuming we're passing zero into the method you exposed). As a result, I would argue for only exposing a method like what I proposed, since I really do think it will cover the vast majority of cases, and even the rare ones have a straightforward solution.

Basically, I'm arguing for trading off some flexibility (passing two arbitrary CGFloats) for a simpler API (just a YES or NO). It also makes it a little harder for magic numbers to get hardcoded, which is good.

Do you think that's reasonable?

**Comment by rahul-malik** [comment]
Yeah I think thats reasonable. I'll update the method I wrote as discussed and assume you'll take care of the fixed size variant.

**Comment by rahul-malik** [comment]
Just updated the code. Let me know what you think.

**Comment by smileyborg** [comment]
Cool, I will take a close look and integrate everything as soon as I get a chance to also work on the other variant.

**Comment by smileyborg** [comment]
Alright, I've gotten your changes merged in along with the corresponding new method for the fixed size variant. Thanks again for contributing!

Please do pull the latest and test both methods out -- make sure they're working for you.
